File: go/wildling/token.go
package wildling

type tokenOptions struct {
	String      *string
	StartLength *int
	EndLength   *int
	Variants    []string
	Src         string
}

type Token struct {
	src         string
	startLength int
	endLength   int
	variants    []string
	count       int
}

func defaultInteger(option *int, fallback int) int {
	if option != nil && *option >= 0 {
		return *option
	}
	return fallback
}

func powInt(base, exp int) int {
	result := 1
	for i := 0; i < exp; i++ {
		result *= base
	}
	return result
}

func newToken(options tokenOptions) *Token {
	startLength := defaultInteger(options.StartLength, 1)
	endLength := defaultInteger(options.EndLength, 1)
	variants := options.Variants
	if variants == nil {
		variants = []string{}
	}

	count := 0
	for length := startLength; length <= endLength; length++ {
		count += powInt(len(variants), length)
	}

	return &Token{
		src:         options.Src,
		startLength: startLength,
		endLength:   endLength,
		variants:    variants,
		count:       count,
	}
}

func (t *Token) Count() int {
	return t.count
}

func (t *Token) Src() string {
	return t.src
}

func (t *Token) Get(index int) string {
	if index > t.count-1 || index < 0 {
		return ""
	}
	if index == 0 && t.startLength == 0 {
		return ""
	}

	indexWithOffset := index
	stringLength := t.startLength
	for stringLength = t.startLength; stringLength <= t.endLength; stringLength++ {
		offsetCount := powInt(len(t.variants), stringLength)
		if indexWithOffset < offsetCount {
			break
		}
		indexWithOffset -= offsetCount
	}

	parts := make([]string, stringLength)
	for i := 0; i < stringLength; i++ {
		variantIndex := indexWithOffset % len(t.variants)
		indexWithOffset /= len(t.variants)
		parts[i] = t.variants[variantIndex]
	}
	out := ""
	for _, p := range parts {
		out += p
	}
	return out
}

File: scala/src/wildling/Token.scala
package wildling

final class TokenOptions(
    var string: Option[String] = None,
    var startLength: Option[Int] = None,
    var endLength: Option[Int] = None,
    var variants: Option[Seq[String]] = None,
    var src: Option[String] = None
)

object TokenOptions {
  def of(
      variants: Seq[String],
      startLength: Int,
      endLength: Int,
      src: String
  ): TokenOptions =
    new TokenOptions(
      startLength = Some(startLength),
      endLength = Some(endLength),
      variants = Some(variants),
      src = Some(src)
    )
}

final class Token(options: TokenOptions) {
  private val src: String = options.src.getOrElse("")
  private val startLength: Int = Token.defaultInteger(options.startLength, 1)
  private val endLength: Int = Token.defaultInteger(options.endLength, 1)
  private val variants: Seq[String] = options.variants.getOrElse(Seq.empty)
  private val countValue: Int = {
    var total = 0
    var length = startLength
    while (length <= endLength) {
      total += Token.pow(variants.size, length)
      length += 1
    }
    total
  }

  def count(): Int = countValue

  def srcValue(): String = src

  def get(index: Int): String = {
    if (index > countValue - 1 || index < 0) return ""
    if (index == 0 && startLength == 0) return ""

    var indexWithOffset = index
    var stringLength = startLength
    var length = startLength
    var done = false
    while (length <= endLength && !done) {
      stringLength = length
      val offsetCount = Token.pow(variants.size, length)
      if (indexWithOffset < offsetCount) {
        done = true
      } else {
        indexWithOffset -= offsetCount
        length += 1
      }
    }

    val out = new StringBuilder
    var i = 0
    while (i < stringLength) {
      val variantIndex = indexWithOffset % variants.size
      indexWithOffset /= variants.size
      out.append(variants(variantIndex))
      i += 1
    }
    out.toString
  }
}

object Token {
  private def defaultInteger(option: Option[Int], fallback: Int): Int =
    option.filter(_ >= 0).getOrElse(fallback)

  private def pow(base: Int, exp: Int): Int = {
    var result = 1
    var i = 0
    while (i < exp) {
      result *= base
      i += 1
    }
    result
  }
}
